fsck/fsck.c

```diff
@@ -670,58 +670,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
 		return -ENOMEM;
 	}
 
+	if (main_sect)
+		bs_offset = BOOT_SEC_IDX;
+	else
+		bs_offset = BACKUP_BOOT_SEC_IDX;
+
 	if (exfat_read(bd->dev_fd, bs, sizeof(*bs),
-			bs_offset * bd->sector_size) != (ssize_t)sizeof(*bs)) {
+			bs_offset * sect_size) != (ssize_t)sizeof(*bs)) {
 		exfat_err("failed to read a boot sector\n");
 		ret = -EIO;
 		goto err;
 	}
 
 	if (memcmp(bs->bpb.oem_name, "EXFAT   ", 8) != 0) {
-		exfat_err("failed to find exfat file system.\n");
+		if (main_sect)
+			exfat_err("failed to find exfat file system\n");
 		goto err;
 	}
 
-	ret = boot_region_checksum(bd, bs_offset);
+	ret = boot_region_checksum(bd, bs_offset, sect_size);
 	if (ret < 0)
 		goto err;
 
 	ret = -EINVAL;
 	if (EXFAT_SECTOR_SIZE(bs) < 512 || EXFAT_SECTOR_SIZE(bs) > 4 * KB) {
-		exfat_err("too small or big sector size: %d\n",
-				EXFAT_SECTOR_SIZE(bs));
+		if (main_sect)
+			exfat_err("too small or big sector size: %d\n",
+				  EXFAT_SECTOR_SIZE(bs));
 		goto err;
 	}
 
```

Comment:
I agreed Eric's suggestion that update bd->sector_size here. I think that's the best way to do it.

Reply:
I think  Eric's suggestion can cause the confusion about device's sector size and Exfat's sector size, But it is the easiest way to solve the problem which comes up when device's sector size and Exfat's sector size are different. So I think we can apply the suggestion right now. But fsck does not need that, because fsck does not depend on the device sector size.

Reply:
if so, we should not use struct bd->dev_fd also. Do you think it is not confused ? we should think exfat image as bd.

Reply:
> because fsck does not depend on the device sector size.
If you used struct bd in fsck, please don't make more complex change, which mean we don't need to separate bd->dev_fd and exfat sector size.

Reply:
Okay, this is simple.